include/particles/particle_bbar.tcc

```diff
@@ -0,0 +1,182 @@
+  }
+
+  if (std::fabs(strain_rate(0)) < 1.E-15) strain_rate[0] = 0.;
+  return strain_rate;
+}
+
+// Compute strain rate of the particle
+template <>
+inline Eigen::Matrix<double, 6, 1> mpm::ParticleBbar<2>::compute_strain_rate(
+    const Eigen::MatrixXd& dn_dx, unsigned phase) noexcept {
+  // Define strain rate
+  Eigen::Matrix<double, 6, 1> strain_rate = Eigen::Matrix<double, 6, 1>::Zero();
+
+  for (unsigned i = 0; i < this->nodes_.size(); ++i) {
+    Eigen::Matrix<double, 2, 1> vel = nodes_[i]->velocity(phase);
+    strain_rate[0] +=
+        (dn_dx(i, 0) + (dn_dx_centroid_(i, 0) - dn_dx(i, 0)) / 3.) * vel[0] +
+        (dn_dx_centroid_(i, 1) - dn_dx(i, 1)) / 3. * vel[1];
+    strain_rate[1] +=
+        (dn_dx_centroid_(i, 0) - dn_dx(i, 0)) / 3. * vel[0] +
+        (dn_dx(i, 1) + (dn_dx_centroid_(i, 1) - dn_dx(i, 1)) / 3.) * vel[1];
+    strain_rate[2] += (dn_dx_centroid_(i, 0) - dn_dx(i, 0)) / 3. * vel[0] +
+                      (dn_dx_centroid_(i, 1) - dn_dx(i, 1)) / 3. * vel[1];
+    strain_rate[3] += dn_dx(i, 1) * vel[0] + dn_dx(i, 0) * vel[1];
+  }
+
+  if (std::fabs(strain_rate[0]) < 1.E-15) strain_rate[0] = 0.;
+  if (std::fabs(strain_rate[1]) < 1.E-15) strain_rate[1] = 0.;
+  if (std::fabs(strain_rate[2]) < 1.E-15) strain_rate[2] = 0.;
+  if (std::fabs(strain_rate[3]) < 1.E-15) strain_rate[3] = 0.;
+  return strain_rate;
+}
+
+// Compute strain rate of the particle
+template <>
+inline Eigen::Matrix<double, 6, 1> mpm::ParticleBbar<3>::compute_strain_rate(
+    const Eigen::MatrixXd& dn_dx, unsigned phase) noexcept {
+  // Define strain rate
+  Eigen::Matrix<double, 6, 1> strain_rate = Eigen::Matrix<double, 6, 1>::Zero();
+
+  for (unsigned i = 0; i < this->nodes_.size(); ++i) {
+    Eigen::Matrix<double, 3, 1> vel = nodes_[i]->velocity(phase);
+    strain_rate[0] +=
+        (dn_dx(i, 0) + (dn_dx_centroid_(i, 0) - dn_dx_(i, 0)) / 3.) * vel[0] +
+        (dn_dx_centroid_(i, 1) - dn_dx_(i, 1)) / 3. * vel[1] +
+        (dn_dx_centroid_(i, 2) - dn_dx_(i, 2)) / 3. * vel[2];
+    strain_rate[1] +=
+        (dn_dx_centroid_(i, 0) - dn_dx_(i, 0)) / 3. * vel[0] +
+        (dn_dx(i, 1) + (dn_dx_centroid_(i, 1) - dn_dx_(i, 1)) / 3.) * vel[1] +
+        (dn_dx_centroid_(i, 2) - dn_dx_(i, 2)) / 3. * vel[2];
+    strain_rate[2] +=
+        (dn_dx_centroid_(i, 0) - dn_dx_(i, 0)) / 3. * vel[0] +
+        (dn_dx_centroid_(i, 1) - dn_dx_(i, 1)) / 3. * vel[1] +
+        (dn_dx(i, 2) + (dn_dx_centroid_(i, 1) - dn_dx(i, 1)) / 3.) * vel[2];
+    strain_rate[3] += dn_dx(i, 1) * vel[0] + dn_dx(i, 0) * vel[1];
+    strain_rate[4] += dn_dx(i, 2) * vel[1] + dn_dx(i, 1) * vel[2];
+    strain_rate[5] += dn_dx(i, 2) * vel[0] + dn_dx(i, 0) * vel[2];
+  }
+
+  for (unsigned i = 0; i < strain_rate.size(); ++i)
+    if (std::fabs(strain_rate[i]) < 1.E-15) strain_rate[i] = 0.;
+  return strain_rate;
+}
+
+// Compute strain of the particle
+template <unsigned Tdim>
+void mpm::ParticleBbar<Tdim>::compute_strain(double dt) noexcept {
```

Comment:
The same with this function. This is the same as the base class. Let's not duplicate.